File: study/src/Patterns/proxy/Plant.java
package Patterns.proxy;

public class Plant extends Planet implements Runnable{
    private double warehouseCapacity = 20;
    private double warehouseLoad = 0;
    private double quantityToProduce = 0;
    private double quantityProduced = 0;
    private double maxBatchSize = 3;
    //TODO move timeForOnePiece into Galaxy as a general parameter
    private long timeForOnePiece = 1000;
    private Mine plantMine;

    public Plant(int x, int y, int radius, int r, int g, int b, Mine mine) {
        super(x, y, radius, r, g, b);
        this.plantMine = mine;
    }

    //TODO create collback for order. Create Customer class
    public boolean placeOrder(double quantity) {
        if (quantityToProduce == 0) {
            this.quantityToProduce = quantity;
            return true;
        }
        return false;
    }

    //TODO to implement production method
    public void produce(double quantityToProduce) {

    }


    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            if(dockedShip != null) {
                if(dockedShip.getLoad() == 0 && quantityToProduce > warehouseLoad) {
                    dockedShip.addPointToRoute(plantMine);
                    dockedShip.addPointToRoute(this);
                    super.undock();
                } else {
                    double batchSize = maxBatchSize;
                    if(dockedShip.getLoad() < maxBatchSize) {
                        batchSize = dockedShip.getLoad();
                    }

                    if(batchSize > warehouseCapacity - warehouseLoad) {
                        batchSize = warehouseCapacity - warehouseLoad;
                    }
                    System.out.println("Docked ship load: " + dockedShip.getLoad());
                    System.out.println("Batch size: " + batchSize);
                    System.out.println("Plant warehouse load: " + warehouseLoad);
                    warehouseLoad += dockedShip.batchUnload(batchSize);
                }
            }

            try {
                Thread.sleep(timeForOnePiece);
            } catch (InterruptedException e) {
                break;
            }
        }

    }
}

File: study/src/kt/proxy_k/Plant.kt
package kt.proxy_k

class Plant(x: Double,
            y: Double,
            radius: Double,
            r: Int,
            g: Int,
            b: Int,

            private var warehouseCapacity: Int = 20,
            private var warehouseLoad: Int = 0,
            private var timeForOnePiece: Long = 1000,
            private var maxBatchSize: Int = 3,
            private var plantMine: Mine,
            private var quantityToProduce: Double = 0.0,
            private var quantityProduced: Double = 0.0
            ) : Planet(x, y, radius, r, g, b), Runnable {

    fun placeOrder(quantity: Double): Boolean {
        if(quantityToProduce.equals(0.0)) {
            quantityToProduce = quantity
            return true
        }
        return false
    }

    override fun run() {
        while (!Thread.currentThread().isInterrupted) {

            dockedShip?.let {
                if (it.getLoad() == 0 && quantityToProduce > warehouseLoad) {
                    it.addPointToRoute(plantMine)
                    it.addPointToRoute(this)
                    undock()
                } else {
                    var batchSize = maxBatchSize
                    if (it.getFreeCapacity() < maxBatchSize) {
                        batchSize = it.getLoad()
                    }

                    if(batchSize > warehouseCapacity - warehouseLoad) {
                        batchSize = warehouseCapacity - warehouseLoad
                    }
                        println("Docked ship load: ${it.getLoad()}")
                        println("Batch size: $batchSize")
                        println("Plant warehouse load: $warehouseLoad")
                        warehouseLoad += it.batchUnload(batchSize)
                }
            }

            try {
                Thread.sleep(timeForOnePiece)
            } catch (e: InterruptedException) {
                break
            }
        }
    }
}